File: 15-3sum/3sum.js
var threeSum = function(nums) {
    let lookup = new Set()
    let output = []
    nums.sort((a, b)=> a - b)

    for (let i = 0; i < nums.length - 2; i++) {
        let j = i + 1
        let k = nums.length - 1

        while (j < k) {
            let sum = nums[i] + nums[j] + nums[k]
            let triplets = nums[i] + "," + nums[j] + "," + nums[k]
            if (sum == 0) {
                if (!lookup.has(triplets)) {
                    lookup.add(triplets)
                    output.push([nums[i], nums[j], nums[k]])
                } 
                j ++
                k --
            } 
            else if (sum < 0) j ++
            else k --
        }
    }

    return output
};

File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        output = []
        lookup = set()
        
        for i in range(len(nums) - 1):
            first = nums[i]

            a = i + 1
            b = len(nums) - 1

            while (a < b):
                sum = first + nums[a] + nums[b]
                if (sum == 0):
                    triplets = (first, nums[a], nums[b])
                    if (triplets not in lookup):
                        output.append([first, nums[a], nums[b]])
                    lookup.add(triplets)
                    a += 1
                    b -= 1
                elif (sum < 0):
                    a += 1
                else:
                    b -= 1

        return output

        